File: lib/java/src/main/java/io/github/telepact/internal/validation/ValidateUnionStruct.java
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package io.github.telepact.internal.validation;

import static io.github.telepact.internal.validation.ValidateStructFields.validateStructFields;

import java.util.List;
import java.util.Map;

import io.github.telepact.internal.types.TStruct;

public class ValidateUnionStruct {
    static List<ValidationFailure> validateUnionStruct(
            TStruct unionStruct,
            String unionTag,
            Map<String, Object> actual, Map<String, Object> selectedTags, ValidateContext ctx) {
        final var selectedFields = selectedTags == null ? null : (List<String>) selectedTags.get(unionTag);

        ctx.path.push(unionTag);

        final var result = validateStructFields(unionStruct.fields, selectedFields, actual, ctx);

        ctx.path.pop();

        return result;
    }
}

File: lib/py/telepact/internal/validation/ValidateUnionStruct.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import TYPE_CHECKING, cast
from ...internal.validation.ValidationFailure import ValidationFailure
from ...internal.validation.ValidateStructFields import validate_struct_fields

if TYPE_CHECKING:
    from ...internal.validation.ValidateContext import ValidateContext
    from ..types.TStruct import TStruct
    from ..types.TTypeDeclaration import TTypeDeclaration


def validate_union_struct(
    union_struct: 'TStruct',
    union_tag: str,
    actual: dict[str, object],
    selected_tags: dict[str, object],
    ctx: 'ValidateContext'
) -> list['ValidationFailure']:
    selected_fields = cast(list[str], selected_tags.get(
        union_tag)) if selected_tags else None

    ctx.path.append(union_tag)

    result = validate_struct_fields(
        union_struct.fields,
        selected_fields,
        actual,
        ctx
    )

    ctx.path.pop()

    return result
